p2panda-group/src/two_party/two_party.rs

```diff
@@ -0,0 +1,750 @@
+                };
+
+                let plaintext = x3dh_decrypt(
+                    &ciphertext,
+                    KEY::identity_secret(&y_manager_i),
+                    KEY::prekey_secret(&y_manager_i),
+                    onetime_secret.as_ref(),
+                )?;
+
+                (y_manager_i, plaintext)
+            }
+            KeyUsed::ReceivedKey => {
+                let TwoPartyCiphertext::Hpke(ciphertext) = ciphertext else {
+                    return Err(TwoPartyError::InvalidCiphertextType);
+                };
+
+                let Some(our_received_secret_key) = &y.our_received_secret_key else {
+                    return Err(TwoPartyError::UnknownSecretUsed(0));
+                };
+
+                let plaintext = hpke_open(&ciphertext, our_received_secret_key, None, None)?;
+                (y_manager, plaintext)
+            }
+            KeyUsed::OwnKey(index) => {
+                let TwoPartyCiphertext::Hpke(ciphertext) = ciphertext else {
+                    return Err(TwoPartyError::InvalidCiphertextType);
+                };
+
+                let plaintext = match y.our_secret_keys.get(&index) {
+                    Some(secret) => hpke_open(&ciphertext, secret, None, None)?,
+                    None => return Err(TwoPartyError::UnknownSecretUsed(index)),
+                };
+
+                for i in y.our_min_key_index..index + 1 {
+                    y.our_secret_keys.remove(&i);
+                }
+                y.our_min_key_index = index + 1;
+
+                (y_manager, plaintext)
+            }
+        };
+
+        Ok((y, y_manager_i, plaintext))
+    }
+}
+
+impl<KEY, KB> TwoParty<KEY, KB> {
+    /// Generate fresh key material for us and the other party for future 2SM rounds.
+    ///
+    /// This material is sent as part of the encrypted ciphertext, attached next to the actual
+    /// secret message. Each party prepares the received keys to be available for future rounds.
+    fn generate_keys(rng: &Rng) -> TwoPartyResult<(NewKeysForUs, NewKeysForThem)> {
+        let our_new_secret = SecretKey::from_bytes(rng.random_array()?);
+        let our_new_public_key = our_new_secret.public_key()?;
+
+        let their_new_secret = SecretKey::from_bytes(rng.random_array()?);
+        let their_new_public_key = their_new_secret.public_key()?;
+
+        Ok((
+            NewKeysForUs {
+                our_new_secret,
+                their_new_public_key,
+            },
+            NewKeysForThem {
+                our_new_public_key,
+                their_new_secret,
+            },
+        ))
+    }
+}
+
+struct NewKeysForUs {
+    our_new_secret: SecretKey,
+    their_new_public_key: PublicKey,
+}
+
+struct NewKeysForThem {
+    our_new_public_key: PublicKey,
+    their_new_secret: SecretKey,
+}
+
+pub type TwoPartyResult<T> = Result<T, TwoPartyError>;
+
+#[derive(Debug, Error)]
+pub enum TwoPartyError {
+    #[error(transparent)]
+    Hpke(#[from] HpkeError),
```

Comment:
For consistency I'd suggest changing `Hpke` and `HpkeError` to `HPKE` and `HPKEError`, both to match `X3DH` and because it is capitalised throughout the existing code and docs.

Reply:
I've changed X3DH to X3dh to make it consistent as this was the only exception :+1: